// @framer/sonnylazuardi.timer-countdown-target-frame/code/TimerCountdown.tsx
import * as React from "react";
import {
  Frame,
  addPropertyControls,
  ControlType,
  //@ts-ignore
  useNavigation,
  RenderTarget
} from "framer";

export function TimerCountdown(props) {
  const { countdown, ...rest } = props;
  const navigation = useNavigation();

  React.useEffect(() => {
    setTimeout(() => {
      if (props.nextFrame) {
        const Component = props.nextFrame[0];
        navigation.push(Component, {
          appearsFrom: "right"
        });
      }
    }, countdown * 1000);
  });
  if (
    RenderTarget.current() === RenderTarget.canvas ||
    RenderTarget.current() === RenderTarget.thumbnail
  ) {
    return (
      <Frame
        {...rest}
        background={"#FEE469"}
        size="100%"
        style={{
          color: "#000",
          fontSize: 16,
          fontWeight: 600
        }}
      >
        {countdown}s
      </Frame>
    );
  } else {
    return null;
  }
}

TimerCountdown.defaultProps = {
  height: 128,
  width: 240,
  countdown: 3
};

addPropertyControls(TimerCountdown, {
  countdown: {
    title: "Countdown",
    type: ControlType.Number
  },
  nextFrame: {
    title: "Next Frame",
    type: ControlType.ComponentInstance
  }
});
